fix [DMAX] some (newer?) videos cannot be downloaded. #14735

w4grfw opened the pull request:
<!--
    **IMPORTANT**: PRs without the template will be CLOSED
    
    Due to the high volume of pull requests, it may be a while before your PR is reviewed.
    Please try to keep your pull request focused on a single bugfix or new feature.
    Pull requests with a vast scope and/or very large diff will take much longer to review.
    It is recommended for new contributors to stick to smaller pull requests, so you can receive much more immediate feedback as you familiarize yourself with the codebase.

    PLEASE AVOID FORCE-PUSHING after opening a PR, as it makes reviewing more difficult.
-->

### Description of your *pull request* and other information

Fixes #14735

based on fix from @dirkf (https://github.com/yt-dlp/yt-dlp/issues/14735#issuecomment-3443241457)
<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--
    # PLEASE FOLLOW THE GUIDE BELOW

    - You will be asked some questions, please read them **carefully** and answer honestly
    - Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
    - Use *Preview* tab to see what your *pull request* will actually look like
-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check those that apply and remove the others:
- [ ] I am the original author of the code in this PR, and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of the code in this PR, but it is in the public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)
- [x] I have read the [policy against AI/LLM contributions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#automated-contributions-ai--llm-policy) and understand I may be blocked from the repository if it is violated

### What is the purpose of your *pull request*? Check those that apply and remove the others:
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

Commit:
fix [DMAX] some (newer?) videos cannot be downloaded. #14735

w4grfw commented on yt_dlp/extractor/dplay.py:
```diff
@@ -1106,14 +1107,37 @@ class DiscoveryNetworksDeIE(DiscoveryPlusBaseIE):
+            'thumbnail': 'https://eu1-prod-images.disco-api.com/2025/10/14/0bdee68c-a8d8-33d9-9204-16eb61108552.jpeg',
+            'tags': [],
+        },
+        'params': {'skip_download': 'm3u8'},
```
I’ve run `hatch test DiscoveryNetworksDeIE`. I’m not sure if it’s sufficient.

w4grfw replied:
The link from the new test does not work with the current release. Should I remove the new test?

bashonly replied:
> The link from the new test does not work with the current release. Should I remove the new test?

It doesn't work with this PR branch either? Then add a `skip` key to the test dict with the reason as the value

w4grfw committed:
Update yt_dlp/extractor/dplay.py

Co-authored-by: bashonly <88596187+bashonly@users.noreply.github.com>

bashonly commented on yt_dlp/extractor/dplay.py:
```diff
@@ -1106,14 +1107,37 @@ class DiscoveryNetworksDeIE(DiscoveryPlusBaseIE):
+        meta = self._download_json(
+            f'https://de-api.loma-cms.com/feloma/videos/{alternate_id}/?environment=dmax&v=2&filter[show.slug]={programme}',
+            alternate_id, fatal=False)
+        video_id = traverse_obj(meta, ('uid', {lambda s: s[-7:]})) or f'{programme}/{alternate_id}'
```
> Older episodes from DMAX, like the link from the first test, do work with the current release. So the second option?

Sounds good

w4grfw committed:
Update yt_dlp/extractor/dplay.py

Co-authored-by: bashonly <88596187+bashonly@users.noreply.github.com>

Commit:
missing keys added and md5-hash fixed

dirkf commented on yt_dlp/extractor/dplay.py:
```diff
@@ -1106,14 +1107,44 @@ class DiscoveryNetworksDeIE(DiscoveryPlusBaseIE):
+                'filter[show.slug]': programme,
+            }, fatal=False)
+        video_id = traverse_obj(meta, ('uid', {str}, {lambda s: s[-7:]})) or f'{programme}/{alternate_id}'
+
```
`meta['taxonomies']` has category/tag data that isn't extracted now for the new shows in `_get_disco_api_info()` despite being requested in the API call there.

The data can be extracted with `traverse_obj(meta, ('taxonomies', lambda _, v: v.get('category') == 'genre', 'title', {strip_or_none}))` and merged into the dict returned by `_get_disco_api_info()`. If the result is set into `categories`, the test values have to be updated, for test 0
```py
            'categories': ['Gold', 'Schatzsucher'],
```
and for new test 1:
```py
            'categories': ['DMAX Originals', 'Jobs', 'Blaulicht'],
```
In case it's disturbing to have the same value in both `categories` and `tags` (eg, `Schatzsucher`/`schatzsucher`), one could decide to merge these into `tags` (eg, `tags.extend(c for c in categories if c.lower() not in tags)`), or to remove any tag value from the `categories`.

w4grfw replied:
The same seems to true for tlc.de

dirkf commented on yt_dlp/extractor/dplay.py:
```diff
@@ -1062,7 +1063,7 @@ class DiscoveryNetworksDeIE(DiscoveryPlusBaseIE):
             'title': 'German Gold',
             'description': 'md5:f3073306553a8d9b40e6ac4cdbf09fc6',
-            'display_id': 'goldrausch-in-australien/german-gold',
+            'display_id': '4756322',
```
As the series/slug is not being passed into `_get_disco_api_info()` for the new case, the `display_id` has either to be re-created there by changing this in the returned dict:
```diff
-            'display_id': display_id,
+            'display_id': '/'.join(traverse_obj(video, (
+                (('included', ...), 'data'), 'attributes', 'alternateId', {strip_or_none}, filter)))
+            if display_id == video_id else display_id,
```
or just re-applied to the returned result (but that would only handle the DMAX case).
 
Then this value should be reverted and the test result for this new test should be have: 
```py
            'display_id': 'feuerwache-3-alarm-in-muenchen/24-stunden-auf-der-feuerwache-3',
```

dirkf reviewed (commented):
Looking good, but these points might also be considered.

Commit:
dplay.co.uk removed because it no longer exists

Commit:
use new api for tlc.de

Commit:
new key categories

Commit:
fallback value for video_id and display_id

as suggested in
https://github.com/yt-dlp/yt-dlp/pull/14818#discussion_r2483809993

bashonly commented on yt_dlp/extractor/dplay.py:
```diff
@@ -1100,20 +1102,93 @@ class DiscoveryNetworksDeIE(DiscoveryPlusBaseIE):
+        disco_api_info['display_id'] = alternate_id
+        disco_api_info['categories'] = traverse_obj(meta, (
+            'taxonomies', lambda _, v: v['category'] == 'genre', 'title', {str.strip}, filter, all, filter))
+        return disco_api_info
```
```suggestion
        video_id = traverse_obj(meta, ('uid', {str}, {lambda s: s[-7:]})) or alternate_id

        disco_api_info = self._get_disco_api_info(
            url, video_id, 'eu1-prod.disco-api.com', domain.replace('.', ''), 'DE')
        disco_api_info['display_id'] = alternate_id
        disco_api_info['categories'] = traverse_obj(meta, (
            'taxonomies', lambda _, v: v['category'] == 'genre', 'title', {str.strip}, filter, all, filter))

        return disco_api_info
```

bashonly committed:
Apply suggestions from code review

bashonly committed:
Update yt_dlp/extractor/dplay.py
